crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs

```diff
@@ -554,11 +617,94 @@ impl GroupSessionManager {
         // needed because each encryption step will mutate the Olm session,
         // ratcheting its state forward.
         for result in join_all(tasks).await {
-            let used_sessions: OlmResult<Vec<Session>> = result.expect("Encryption task panicked");
+            let result: OlmResult<(Vec<Session>, BTreeMap<OwnedUserId, Vec<OwnedDeviceId>>)> =
+                result.expect("Encryption task panicked");
 
-            changes.sessions.extend(used_sessions?);
+            let (used_sessions, failed_no_olm) = result?;
+            changes.sessions.extend(used_sessions);
+
+            no_olm.extend(failed_no_olm.into_iter())
         }
 
+        // Handle unverified and blacklisted withhelds
+        let mut flat_with_code: Vec<(OwnedUserId, OwnedDeviceId, WithheldCode)> = withheld_devices
+            .into_iter()
+            .flat_map(|(_, list)| {
+                list.into_iter().map(|(d, c)| (d.user_id().to_owned(), d.device_id().to_owned(), c))
+            })
+            .collect();
+
+        no_olm.iter().for_each(|(u, list)| {
+            list.into_iter().for_each(|d| {
+                flat_with_code.push((u.to_owned(), d.to_owned(), WithheldCode::NoOlm))
+            });
+        });
+
+        let withheld_to_device_request: Vec<Arc<ToDeviceRequest>> = flat_with_code
+            .chunks(Self::MAX_TO_DEVICE_MESSAGES)
+            .map(|chunk| {
+                let mut messages = BTreeMap::new();
+
+                // TODO do not resend code if already sent for that session (or target if
+                // no_olm)
+                chunk.into_iter().for_each(|(user_id, device_id, code)| {
+                    // TODO use helper to create correct content depending on alg and code
+                    let content = MegolmV1AesSha2WithheldContent::new(
+                        room_id.to_owned(),
+                        Some(outbound.session_id().to_owned()),
+                        // help: can't access it from the outbound, should I get it from Inbound?
+                        self.account.identity_keys.curve25519,
+                        code.to_owned(),
+                        // help: can't access it from the outbound, should I get it from Inbound?
+                        Some(self.account.device_id.deref().to_owned()),
+                    );
+                    let content = Raw::new(&content)
+                        .expect("We can always serialize a withheld content info")
+                        .cast();
+                    messages
+                        .entry(user_id.to_owned())
+                        .or_insert_with(BTreeMap::new)
+                        .insert(DeviceIdOrAllDevices::DeviceId(device_id.to_owned()), content);
+                });
+
+                Arc::new(ToDeviceRequest {
```

Comment:
These requests need to be persisted with the `OutboundGroupSession` am I missing something or is this not done anywhere?

Reply:
done now